File: source/Mlos.NetCore/Security.Windows.cs
// -----------------------------------------------------------------------
// <copyright file="Security.Windows.cs" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root
// for license information.
// </copyright>
// -----------------------------------------------------------------------

using System;
using System.ComponentModel;
using System.Runtime.InteropServices;

namespace Mlos.Core.Windows
{
    /// <summary>
    /// Security helper class.
    /// </summary>
    /// <remarks>
    /// Windows only.
    /// </remarks>
    internal static class Security
    {
        /// <summary>
        /// Gets the current user security identifier.
        /// </summary>
        /// <returns></returns>
        internal static SecurityIdentifierSafePtr GetCurrentUserSid()
        {
            using ProcessTokenSafeHandle currentProcessHandle = Native.GetCurrentProcess();

            if (!Native.OpenProcessToken(currentProcessHandle, Native.AccessRights.TokenQuery, out AccessTokenSafeHandle tokenHandle))
            {
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            using AccessTokenSafeHandle scopedTokenHandle = tokenHandle;

            // Get the token user.
            //
            if (!Native.GetTokenInformation(tokenHandle, Native.TokenInformationClass.TokenUser, IntPtr.Zero, 0, out uint returnLength))
            {
                if (Marshal.GetLastWin32Error() != Native.ErrorInsufficientBuffer)
                {
                    throw new Win32Exception(Marshal.GetLastWin32Error());
                }
            }

            using LocalAllocSafePtr tokenUserPtr = Native.LocalAlloc(Native.LocalMemoryFlags.Fixed | Native.LocalMemoryFlags.ZeroInit, returnLength);
            if (tokenUserPtr.IsInvalid)
            {
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            if (!Native.GetTokenInformation(tokenHandle, Native.TokenInformationClass.TokenUser, tokenUserPtr.DangerousGetHandle(), returnLength, out returnLength))
            {
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            Native.TOKEN_USER tokenUser = Marshal.PtrToStructure<Native.TOKEN_USER>(tokenUserPtr.DangerousGetHandle());

            uint userSidLength = Native.GetLengthSid(tokenUser.User.Sid);

            SecurityIdentifierSafePtr currentUserSidPtr = Native.AllocSecurityIdentifier(Native.LocalMemoryFlags.Fixed | Native.LocalMemoryFlags.ZeroInit, userSidLength);
            if (currentUserSidPtr.IsInvalid)
            {
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            if (!Native.CopySid(userSidLength, currentUserSidPtr, tokenUser.User.Sid))
            {
                currentUserSidPtr.Dispose();
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            return currentUserSidPtr;
        }

        /// <summary>
        /// Gets the current user security identifier as a string.
        /// </summary>
        internal static string CurrentUserSidString
        {
            get
            {
                using (SecurityIdentifierSafePtr currentUserSid = GetCurrentUserSid())
                {
                    if (!Native.ConvertSidToStringSid(currentUserSid.DangerousGetHandle(), out string currentUserSidString))
                    {
                        throw new Win32Exception(Marshal.GetLastWin32Error());
                    }

                    return currentUserSidString;
                }
            }
        }

        /// <summary>
        /// Creates a default security descriptor.
        /// </summary>
        /// <returns></returns>
        internal static SecurityDescriptorSafePtr CreateDefaultSecurityDescriptor()
        {
            string currentUserSid = CurrentUserSidString;

            // Create a security descriptor.
            // General Access for current user.
            // https://itconnect.uw.edu/wares/msinf/other-help/understanding-sddl-syntax/
            //
            string securityDescriptorStr = $"D:P(A;;GA;;;{currentUserSid})";

            return CreateSecurityDescriptorFromString(securityDescriptorStr);
        }

        /// <summary>
        /// Creates a security descriptor from the security descriptor string.
        /// </summary>
        /// <param name="securityDescriptorString"></param>
        /// <returns></returns>
        internal static SecurityDescriptorSafePtr CreateSecurityDescriptorFromString(string securityDescriptorString)
        {
            if (!Native.ConvertStringSecurityDescriptorToSecurityDescriptor(
                securityDescriptorString,
                stringSDRevision: 1,
                out SecurityDescriptorSafePtr securityDescriptor,
                out int _))
            {
                throw new ArgumentException("Invalid security descriptor string", new Win32Exception(Marshal.GetLastWin32Error()));
            }

            return securityDescriptor;
        }

        /// <summary>
        /// Check if handle has been created by authorized user.
        /// </summary>
        /// <param name="handle"></param>
        /// <remarks>
        /// Authorized users:
        ///  - built in administrators,
        ///  - current user.
        /// </remarks>
        internal static void VerifyHandleOwner(SafeHandle handle)
        {
            // Get the security descriptor for given handle.
            //
            int result = Native.GetSecurityInfo(
                handle,
                Native.SecurityObjectType.SecurityFileObject,
                Native.SecurityInformation.OwnerSecurityInformation,
                sidOwner: IntPtr.Zero,
                sidGroup: IntPtr.Zero,
                dacl: IntPtr.Zero,
                sacl: IntPtr.Zero,
                out SecurityDescriptorSafePtr handleSecurityDescriptor);
            using SecurityDescriptorSafePtr scopedHandleSecurityDescriptor = handleSecurityDescriptor;
            if (result != Native.ErrorSuccess)
            {
                throw new Win32Exception(result);
            }

            if (!Native.GetSecurityDescriptorOwner(handleSecurityDescriptor, out IntPtr handleOwnerSid, out bool ownerDefaulted1))
            {
                throw new Win32Exception(Marshal.GetLastWin32Error());
            }

            // Allow access, if owner is local system.
            //
            if (Native.IsWellKnownSid(handleOwnerSid, Native.WellKnownSidType.WinLocalSystemSid))
            {
                return;
            }

            // Allow access, if owner is builtin administator.
            //
            if (Native.IsWellKnownSid(handleOwnerSid, Native.WellKnownSidType.WinBuiltinAdministratorsSid))
            {
                return;
            }

            // Allow acces, if current user is the owner.
            //
            using SecurityIdentifierSafePtr currentOwnerSecuryIdentifier = GetCurrentUserSid();
            if (Native.EqualSid(currentOwnerSecuryIdentifier.DangerousGetHandle(), handleOwnerSid))
            {
                return;
            }

            throw new UnauthorizedAccessException();
        }
    }
}

File: source/Mlos.Core/Security.Windows.cpp
//*********************************************************************
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root
// for license information.
//
// @File: Security.Windows.cpp
//
// Purpose:
//      <description>
//
// Notes:
//      <special-instructions>
//
//*********************************************************************

#include "Mlos.Core.h"

namespace Mlos
{
namespace Core
{
//----------------------------------------------------------------------------
// NAME: Security::GetCurrentUserSecurityIdentifierString
//
// PURPOSE:
//  Gets the current user security identifier.
//
// RETURNS:
//  HRESULT.
//
_Check_return_
HRESULT Security::GetCurrentUserSid(_Out_ PSID& currentUserSid)
{
    HRESULT hr = S_OK;
    HANDLE tokenHandle = INVALID_HANDLE_VALUE;
    PTOKEN_USER tokenUser = nullptr;
    DWORD currentUserSidLength = 0;

    currentUserSid = nullptr;

    if (!OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &tokenHandle))
    {
        hr = HRESULT_FROM_WIN32(GetLastError());
    }

    // Obtain the size of the user information from the token.
    //
    DWORD tokenSize = 0;

    if (SUCCEEDED(hr))
    {
        if (!GetTokenInformation(tokenHandle, TokenUser, nullptr, 0, &tokenSize))
        {
            DWORD lastError = GetLastError();
            if (lastError != ERROR_INSUFFICIENT_BUFFER)
            {
                hr = HRESULT_FROM_WIN32(lastError);
            }
        }
    }

    // Allocate the memory for user token.
    //
    if (SUCCEEDED(hr))
    {
        tokenUser = reinterpret_cast<PTOKEN_USER>(LocalAlloc(LMEM_FIXED | LMEM_ZEROINIT, tokenSize));
        if (tokenUser == nullptr)
        {
            hr = E_OUTOFMEMORY;
        }
    }

    // Retrieve the user information from the token.
    //
    if (SUCCEEDED(hr))
    {
        if (!GetTokenInformation(tokenHandle, TokenUser, tokenUser, tokenSize, &tokenSize))
        {
            hr = HRESULT_FROM_WIN32(GetLastError());
        }
    }

    // Convert user security identifier to string format.
    //
    if (SUCCEEDED(hr))
    {
        currentUserSidLength = GetLengthSid(tokenUser->User.Sid);

        currentUserSid = reinterpret_cast<PSID>(LocalAlloc(LMEM_FIXED | LMEM_ZEROINIT, currentUserSidLength));
        if (currentUserSid == nullptr)
        {
            hr = E_OUTOFMEMORY;
        }
    }

    if (SUCCEEDED(hr))
    {
        if (!CopySid(currentUserSidLength, currentUserSid, tokenUser->User.Sid))
        {
            hr = HRESULT_FROM_WIN32(GetLastError());
        }
    }

    LocalFree(tokenUser);
    CloseHandle(tokenHandle);

    return hr;
}

//----------------------------------------------------------------------------
// NAME: Security::GetCurrentUserSecurityIdentifierString
//
// PURPOSE:
//  Gets the current user security identifier as a string.
//
// RETURNS:
//  HRESULT.
//
_Check_return_
HRESULT Security::GetCurrentUserSidString(_Out_ wchar_t*& currentUserSidString)
{
    HRESULT hr = S_OK;
    PSID currentUserSid = nullptr;

    currentUserSidString = nullptr;

    hr = GetCurrentUserSid(currentUserSid);

    // Convert user security identifier to string format.
    //
    if (SUCCEEDED(hr))
    {
        if (!ConvertSidToStringSidW(currentUserSid, &currentUserSidString))
        {
            hr = HRESULT_FROM_WIN32(GetLastError());
        }
    }

    return hr;
}

//----------------------------------------------------------------------------
// NAME: Security::CreateDefaultSecurityDescriptor
//
// PURPOSE:
//  Creates a default security descriptor.
//
// RETURNS:
//  HRESULT.
//
_Check_return_
HRESULT Security::CreateDefaultSecurityDescriptor(_Out_ PSECURITY_DESCRIPTOR& securityDescriptor)
{
    const int maxSDDLStringLength = 1024;

    LPWSTR currentUserSidString = nullptr;
    WCHAR securityDescriptorString[maxSDDLStringLength] = { 0 };

    HRESULT hr = Security::GetCurrentUserSidString(currentUserSidString);

    if (SUCCEEDED(hr))
    {
        // Create a security descriptor.
        // General Access for current user.
        // https://itconnect.uw.edu/wares/msinf/other-help/understanding-sddl-syntax/
        //
        hr = StringCchPrintfW(
            securityDescriptorString,
            _countof(securityDescriptorString),
            L"D:P(A;;GA;;;%s)",
            currentUserSidString);
    }

    if (SUCCEEDED(hr))
    {
        hr = CreateSecurityDescriptorFromString(securityDescriptorString, securityDescriptor);
    }

    LocalFree(currentUserSidString);

    return hr;
}

//----------------------------------------------------------------------------
// NAME: Security::CreateSecurityDescriptorFromString
//
// PURPOSE:
//  Creates a security descriptor from the security descriptor string.
//
// RETURNS:
//  HRESULT.
//
_Check_return_
HRESULT Security::CreateSecurityDescriptorFromString(
    _In_ const wchar_t* const securityDescriptorString,
    _Out_ PSECURITY_DESCRIPTOR& securityDescriptor)
{
    HRESULT hr = S_OK;

    if (!ConvertStringSecurityDescriptorToSecurityDescriptorW(
        securityDescriptorString,
        SDDL_REVISION_1 /* StringSDRevision */,
        &securityDescriptor,
        nullptr /* SecurityDescriptorSize */))
    {
        hr = HRESULT_FROM_WIN32(GetLastError());
    }

    return hr;
}

//----------------------------------------------------------------------------
// NAME: Security::VerifyHandleOwner
//
// PURPOSE:
//  Checks if handle has been created by authorized user.
//
// Authorized users:
//  - local system,
//  - built in administrators,
//  - current user.
//
// RETURNS:
//  HRESULT.
//
_Check_return_
HRESULT Security::VerifyHandleOwner(HANDLE handle)
{
    PSECURITY_DESCRIPTOR securityDescriptor = nullptr;
    PSID handleOwnerSid = nullptr;
    PSID currentUserSid = nullptr;
    HRESULT hr = S_OK;

    bool isAuthorized = false;

    // Get the security descriptor for given handle.
    //
    int result = GetSecurityInfo(
        handle,
        SE_FILE_OBJECT,
        OWNER_SECURITY_INFORMATION,
        &handleOwnerSid /* sidOwner */,
        nullptr /* sidGroup */,
        nullptr /* dacl */,
        nullptr /* sacl */,
        &securityDescriptor);
    if (result != ERROR_SUCCESS)
    {
        hr = HRESULT_FROM_WIN32(result);
    }

    if (!isAuthorized && SUCCEEDED(hr))
    {
        isAuthorized = IsWellKnownSid(handleOwnerSid, WELL_KNOWN_SID_TYPE::WinLocalSystemSid);
    }

    if (!isAuthorized && SUCCEEDED(hr))
    {
        isAuthorized = IsWellKnownSid(handleOwnerSid, WELL_KNOWN_SID_TYPE::WinBuiltinAdministratorsSid);
    }

    if (!isAuthorized && SUCCEEDED(hr))
    {
        hr = GetCurrentUserSid(currentUserSid);
    }

    if (!isAuthorized && SUCCEEDED(hr))
    {
        isAuthorized = EqualSid(handleOwnerSid, currentUserSid);
    }

    // Free allocated memory.
    //
    LocalFree(securityDescriptor);
    LocalFree(currentUserSid);

    if (!isAuthorized && SUCCEEDED(hr))
    {
        hr = E_ACCESSDENIED;
    }

    return hr;
}
}
}
